### PR TITLE
Fix sending out of order `POSITION` over replication

### PR DESCRIPTION
**Pull request by erikjohnston**
If a worker reconnects to Redis we send out the current positions of all our streams. However, if we're also trying to send out a backlog of RDATA at the same time then we can end up sending a `POSITION` with the current token *before* we've sent all the RDATA before the current token.

This doesn't cause actual bugs as the receiving servers see the POSITION, fetch the relevant rows from the DB, and then ignore the old RDATA as they come in. However, this is inefficient so it'd be better if we didn't  send out-of-order positions

### CODE REVIEW

File: synapse/replication/tcp/handler.py [hunk 1]
```diff
@@ -397,29 +402,23 @@ def get_streams_to_replicate(self) -> List[Stream]:
...
-                )
-            )
+        self._should_announce_positions = True
+        self._notifier.notify_replication()
+
+    def should_announce_positions(self) -> bool:
+        """Check if we should send POSITION commands for all streams ASAP."""
+        return self._should_announce_positions
+
+    def will_should_announce_positions(self) -> None:
```
Comment (clokep) [inline synapse/replication/tcp/handler.py:419]: "will_should" seems confusing?

File: synapse/replication/tcp/handler.py [hunk 2]
```diff
@@ -397,29 +402,23 @@ def get_streams_to_replicate(self) -> List[Stream]:
...
-                    current_token,
-                    current_token,
-                )
-            )
+        self._should_announce_positions = True
+        self._notifier.notify_replication()
+
+    def should_announce_positions(self) -> bool:
+        """Check if we should send POSITION commands for all streams ASAP."""
+        return self._should_announce_positions
```
Comment (clokep) [inline synapse/replication/tcp/handler.py:417]: Maybe just make the attribute public?

File: synapse/replication/tcp/resource.py [hunk 3]
```diff
@@ -158,6 +158,21 @@ async def _run_notifier_loop(self) -> None:
                         all_streams = list(all_streams)
                         random.shuffle(all_streams)
 
+                    if self.command_handler.should_announce_positions():
+                        # We need to send out POSITIONs for all streams, usually
+                        # because a worker has reconnected.
+                        self.command_handler.will_should_announce_positions()
```
Comment (clokep) [inline synapse/replication/tcp/resource.py:164]: `will_should_announce_positions` sets `_should_announce_positions` to false, so we're essentially saying that we've now announced positions, don't do it again next time around the loop?

Is it possible for another `REPLICATE` to come in after this? I guess that's fine, we'll just send them again next loop?

Comment (erikjohnston) [reply synapse/replication/tcp/resource.py:164]: Yup, pretty much. 

> so we're essentially saying that we've now announced positions, don't do it again next time around the loop?

We say we are *about* to send positions (to handle the race where something asks us to send out positions after we've just sent them out but before we clear the flag)

File: synapse/replication/tcp/handler.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:419]: Oops

File: synapse/replication/tcp/handler.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:417]: Yeah, maybe easier. I'm oscillating between whether having the functions there helps formalise how this is meant to work

**Review by clokep — APPROVED** [review]

File: synapse/replication/tcp/handler.py (hunk 2 above)
Comment (clokep) [reply synapse/replication/tcp/handler.py:417]: I don't have strong opinions. 👍

**Pull request merged**